llama_index/vector_stores/google/generativeai/genai_extension.py

```diff
@@ -0,0 +1,534 @@
+                ),
+            )
+            for text, metadata in zip(texts, metadatas)
+        ]
+    )
+)
+
+while not operation.done():
+    _logger.info("Waiting on batch_create_chunk operation")
+response = cast(genai.BatchCreateChunksResponse, operation.result())
+return [chunk for chunk in response.chunks]
+"""
+
+
+def delete_chunk(
+    *,
+    corpus_id: str,
+    document_id: str,
+    chunk_id: str,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> None:
+    if client is None:
+        client = _default_retriever
+    client.delete_chunk(
+        genai.DeleteChunkRequest(
+            name=str(
+                EntityName(
+                    corpus_id=corpus_id, document_id=document_id, chunk_id=chunk_id
+                )
+            )
+        )
+    )
+
+
+def query_corpus(
+    *,
+    corpus_id: str,
+    query: str,
+    k: int = 4,
+    filter: Optional[Dict[str, Any]] = None,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> List[genai.RelevantChunk]:
+    if client is None:
+        client = _default_retriever
+    response = client.query_corpus(
+        genai.QueryCorpusRequest(
+            name=str(EntityName(corpus_id=corpus_id)),
+            query=query,
+            metadata_filters=_convertFilter(filter),
+            results_count=k,
+        )
+    )
+    return list(response.relevant_chunks)
+
+
+def query_document(
+    *,
+    corpus_id: str,
+    document_id: str,
+    query: str,
+    k: int = 4,
+    filter: Optional[Dict[str, Any]] = None,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> List[genai.RelevantChunk]:
+    if client is None:
+        client = _default_retriever
+    response = client.query_document(
+        genai.QueryDocumentRequest(
+            name=str(EntityName(corpus_id=corpus_id, document_id=document_id)),
+            query=query,
+            metadata_filters=_convertFilter(filter),
+            results_count=k,
+        )
+    )
+    return list(response.relevant_chunks)
+
+
+@dataclass
+class Passage:
+    text: str
+    ids: List[str]
+
+
+@dataclass
+class TextAnswer:
+    answer: str
+    attributed_passages: List[Passage]
+    answerable_probability: float
+
+
+def generate_text_answer(
+    *,
+    prompt: str,
+    passages: List[str],
+    answer_style: genai.AnswerStyle,
+    client: Optional[genai.TextServiceClient] = None,
+) -> TextAnswer:
+    if client is None:
+        client = _default_text_service
+    response = client.generate_text_answer(
+        genai.GenerateTextAnswerRequest(
+            question=genai.TextPrompt(text=prompt),
+            model=_default_text_service_model,
+            answer_style=answer_style,
+            grounding_source=genai.GroundingSource(
+                passages=genai.InlinePassages(
+                    passages=[
+                        genai.InlinePassage(
+                            text=chunk,
+                            id=str(uuid.uuid4()),
```

Comment:
nit: these could also just be incremented number, eg "001", "002", etc.

Reply:
I don't have a strong opinion here. I originally thought UUIDs were longer which would eat up context window space, but I remembered we actually convert to sequential ID numbers before sending to the LLM.